pkg/resources/provisioning.cattle.io/v1/cluster/validator_test.go

```diff
@@ -200,3 +202,337 @@ func TestValidateMachinePoolName(t *testing.T) {
+		oldCluster    *v1.Cluster
+		shouldSucceed bool
+	}{
+		{
+			name:          "create no var",
+			request:       &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Create}},
+			cluster:       &v1.Cluster{},
+			oldCluster:    nil,
+			shouldSucceed: true,
+		},
+		{
+			name:    "create with var",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Create}},
+			cluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{
+						{
+							Name:  "CATTLE_AGENT_VAR_DIR",
+							Value: "a",
+						},
+					},
+				},
+			},
+			oldCluster:    nil,
+			shouldSucceed: false,
+		},
+		{
+			name:    "update equal",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{
+						{
+							Name:  "CATTLE_AGENT_VAR_DIR",
+							Value: "a",
+						},
+					},
+				},
+			},
+			oldCluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{
+						{
+							Name:  "CATTLE_AGENT_VAR_DIR",
+							Value: "a",
+						},
+					},
+				},
+			},
+			shouldSucceed: true,
+		},
+		{
+			name:    "update removing var",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig:    &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{},
+				},
+			},
+			oldCluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{
+						{
+							Name:  "CATTLE_AGENT_VAR_DIR",
+							Value: "a",
+						},
+					},
+				},
+			},
+			shouldSucceed: true,
+		},
+		{
+			name:    "update adding var",
+			request: &admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Operation: admissionv1.Update}},
+			cluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig: &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{
+						{
+							Name:  "CATTLE_AGENT_VAR_DIR",
+							Value: "a",
+						},
+					},
+				},
+			},
+			oldCluster: &v1.Cluster{
+				Spec: v1.ClusterSpec{
+					RKEConfig:    &v1.RKEConfig{},
+					AgentEnvVars: []rkev1.EnvVar{},
+				},
+			},
+			shouldSucceed: false,
+		},
+	}
+
+	a := provisioningAdmitter{}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			status := a.validateAgentEnvVars(tt.request, tt.oldCluster, tt.cluster)
+			assert.Equal(t, tt.shouldSucceed, status == nil)
+		})
+	}
+}
+
+func TestValidateDataDirectories(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name          string
+		request       *admission.Request
+		cluster       *v1.Cluster
+		oldCluster    *v1.Cluster
+		shouldSucceed bool
+	}{
+		{
```

Comment:
One other case that would be good to add: the cluster exists, but RKEConfig is nil.